zkevm-circuits/src/evm_circuit/util/memory_gadget.rs

```diff
@@ -1,57 +1,166 @@
 
-    pub(crate) fn value<F: FieldExt>(address: [u8; 32]) -> Address {
-        from_bytes::value::<F>(&address[0..NUM_ADDRESS_BYTES_USED])
-            .get_lower_128() as Address
+    pub(crate) fn value<F: FieldExt>(address: [u8; 32]) -> u64 {
+        from_bytes::value::<F>(&address[..N_BYTES_MEMORY_ADDRESS])
+            .get_lower_128() as u64
     }
 }
 
 /// The sum of bytes of the address that are unused for most calculations on the
 /// address
 pub(crate) mod address_high {
     use crate::evm_circuit::{
-        param::NUM_ADDRESS_BYTES_USED,
+        param::N_BYTES_MEMORY_ADDRESS,
         util::{sum, Word},
     };
     use halo2::{arithmetic::FieldExt, plonk::Expression};
 
     pub(crate) fn expr<F: FieldExt>(address: &Word<F>) -> Expression<F> {
-        sum::expr(&address.cells[NUM_ADDRESS_BYTES_USED..32])
+        sum::expr(&address.cells[N_BYTES_MEMORY_ADDRESS..])
     }
 
     pub(crate) fn value<F: FieldExt>(address: [u8; 32]) -> F {
-        sum::value::<F>(&address[NUM_ADDRESS_BYTES_USED..32])
+        sum::value::<F>(&address[N_BYTES_MEMORY_ADDRESS..])
+    }
+}
+
+/// Convert the dynamic memory offset and length from random linear combiation
+/// to integer. It handles the "no expansion" feature when length is zero.
+#[derive(Clone, Debug)]
+pub(crate) struct MemoryAddressGadget<F> {
+    memory_offset: Cell<F>,
+    memory_offset_bytes: MemoryAddress<F>,
+    memory_length: MemoryAddress<F>,
+    memory_length_is_zero: IsZeroGadget<F>,
+}
+
+impl<F: FieldExt> MemoryAddressGadget<F> {
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        memory_offset: Cell<F>,
+        memory_length: MemoryAddress<F>,
+    ) -> Self {
+        let memory_length_is_zero =
+            IsZeroGadget::construct(cb, sum::expr(&memory_length.cells));
+        let memory_offset_bytes = cb.query_rlc();
+
+        let has_length = 1.expr() - memory_length_is_zero.expr();
+        cb.condition(has_length, |cb| {
+            cb.require_equal(
+                "Offset decomposition into 5 bytes",
+                memory_offset_bytes.expr(),
+                memory_offset.expr(),
+            );
+        });
+
+        Self {
+            memory_offset,
+            memory_offset_bytes,
+            memory_length,
+            memory_length_is_zero,
+        }
+    }
+
+    pub(crate) fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        memory_offset: U256,
+        memory_length: U256,
+        randomness: F,
+    ) -> Result<u64, Error> {
+        let memory_offset_bytes = memory_offset.to_le_bytes();
+        let memory_length_bytes = memory_length.to_le_bytes();
+        let memory_length_is_zero = memory_length.is_zero();
+        self.memory_offset.assign(
+            region,
+            offset,
+            Some(Word::random_linear_combine(memory_offset_bytes, randomness)),
+        )?;
+        self.memory_offset_bytes.assign(
+            region,
+            offset,
```

Comment:
While all of the offsets in the same region contain the same value? Is everything done within the same origin row or something similar?

Reply:
The `offset` here is "offset of the step", and the column and rotation information is stored in `Cell` itself, to assign to correct `offset` of its value.